app/models/user.rb

```diff
@@ -5,13 +5,39 @@ class User < ApplicationRecord
   # :confirmable, :lockable, :timeoutable, :trackable and :omniauthable
   devise :database_authenticatable, :registerable,
          :recoverable, :rememberable, :validatable,
-         :confirmable
+         :confirmable, :omniauthable, omniauth_providers: [:google_oauth2]
 
   # the authenticate method from devise documentation
   def self.authenticate(email, password)
     user = User.find_for_authentication(email: email)
     user&.valid_password?(password) ? user : nil
   end
 
+  def self.from_omniauth(auth)
+    case auth.provider
+    when :email
+      return
+    end
+    find_existing_user(auth) || create_new_user(auth)
+  end
+
   has_many :keywords, inverse_of: :user, dependent: :destroy
+  enum provider: { email: 'email', google_oauth2: 'google_oauth2' }
+
+  private_class_method def self.find_existing_user(auth)
```

Comment:
```suggestion
  private_class_method def self.find_by_auth(auth)
```

As ActiveRecord already exposes the "find" method that does "find existing users" (on the User class), we might want a method name that is a little bit more specific to "auth". What about "find_by_auth"?

Reply:
Rubocop auto correct does not like `find_by_auth` I used `find_with_auth` instead.

Reply:
The reason is to avoid codding custom methods that handle exactly the same result as `find_by(auth: auth)`. In our case it is different. Ok with the renaming `with` :harold: